Fixed #25845 -- Fixed incorrect timezone warnings in custom admin templates.

Strayer opened the pull request:
Fixed false timezone warnings if a custom admin base.html template doesn't specify the "data-admin-utc-offset" body attribute.

Added Selenium tests for Firefox, Chrome and IE based on `DateTimePickerShortcutsSeleniumFirefoxTests`, as suggested by @timgraham in #5769.

https://code.djangoproject.com/ticket/25845

timgraham commented on django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:
```diff
@@ -19,9 +19,8 @@
         init: function() {
-            var body = document.getElementsByTagName('body')[0];
-            var serverOffset = body.getAttribute('data-admin-utc-offset');
-            if (serverOffset !== undefined) {
```
what about `body.getAttribute('data-admin-utc-offset') || null`?

Strayer replied:
Should work too and seems more fitting than the extra method. I'll implement this in both checks and remove the serverOffset method.

timgraham commented on tests/admin_widgets/false_timezone_warning_templates/admin/base.html:
```diff
@@ -0,0 +1,88 @@
+{% load i18n admin_static %}<!DOCTYPE html>
```
Try to minimize the template as much as possible to only include the bits necessary for the test case. Also include an explanatory comment like "Intentionally missing data-.... in <body> tag."

Strayer replied:
I could minimize it as much as this:
```
<html>
<head>
{% block extrahead %}{% endblock %}
</head>
<body>
{% block content %}
{{ content }}
{% endblock %}
</body>
</html>
```

But I fear that the test may quickly return false positives because of other issues causing the JavaScript to not run - in that case, the test would return as OK because there are no `.timezonewarning` elements. An example would be if the Django admin at some point doesn't put the JavaScript in `extrahead`, but in the page footer. This can happen with the full file too, of course, but hopefully not as quickly.

One could add a second test case that tests that the warnings are indeed shown with mostly the same template, except that it includes the data attribute. Something like `test_false_timezone_warnings_safety_check`... not sure if this makes sense though.

Would this comment on the top of the file be correct?
`{# Intentionally missing data-admin-utc-offset in body tag to test for false timezone warnings. #}`

timgraham replied:
Not sure -- maybe such a brittle test isn't worthwhile. We could add a comment in DateTimeShortcuts.js instead. I feel like this fix is probably only temporarily important anyway, i.e. going forward people overriding the template won't intentionally omit the attribute.

Strayer replied:
A very simple test would be a plain JavaScript test that ensures that `DateTimeShortcuts.timezoneOffset` is still `0` after running `DateTimeShortcuts.init()`. [This property is checked to determine if the timezone warnings should be shown](https://github.com/django/django/blob/master/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js#L62). This could be done in [js_tests/admin/DateTimeShortcuts.test.js](https://github.com/django/django/blob/master/js_tests/admin/DateTimeShortcuts.test.js) rather easy. Its not really a test if the incorrect warnings are shown, but at least it tests the condition for the timezone warnings. This will also probably cause false positives if the JS test client is running in UTC, unless I find a way to reliably mock this.

timgraham replied:
If the JavaScript test is feasible, that would be simpler.

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -898,6 +899,54 @@ class DateTimePickerAltTimezoneSeleniumIETests(DateTimePickerAltTimezoneSelenium
+        """
+        self.admin_login(username='super', password='secret', login_url='/')
+
+        self.selenium.get('%s%s' % (self.live_server_url,
```
Please use longer lines (we prefer up to 119 chars when it helps readability)
Also, I think the blank lines in the test could be omitted.

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -898,6 +899,54 @@ class DateTimePickerAltTimezoneSeleniumIETests(DateTimePickerAltTimezoneSelenium
+    def test_false_timezone_warnings(self):
+        """
+        Ensure that no false timezone warnings appear if the utc offset is
+        missing in the admin body tag.
```
"tag (#25845)."

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -898,6 +899,54 @@ class DateTimePickerAltTimezoneSeleniumIETests(DateTimePickerAltTimezoneSelenium
 
 
 @override_settings(PASSWORD_HASHERS=['django.contrib.auth.hashers.SHA1PasswordHasher'],
```
you can put PASSWORD_HASHERS on the next line here (the existing style is not preferred).

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -898,6 +899,54 @@ class DateTimePickerAltTimezoneSeleniumIETests(DateTimePickerAltTimezoneSelenium
 
 @override_settings(PASSWORD_HASHERS=['django.contrib.auth.hashers.SHA1PasswordHasher'],
+    ROOT_URLCONF='admin_widgets.urls',
+    TIME_ZONE='America/Chicago',
```
is this important for the test?

Strayer replied:
I thought about this when writing the patch and decided it makes sense to include it in case the timezone for the admin tests change at some point. The incorrect warnings appear as soon as the client browser is not in UTC (and thus has a localOffset), so this line is actually not really useful and I should remove it.

It would make more sense to set the webdrivers timezone, but I couldn't find an easy way to do that without including external JavaScript mock libraries.

It seems like the current test would return a false positive if the webdrivers profile is using the UTC timezone.

timgraham commented:
Please add a mention in docs/releases/1.9.1.txt too.

Strayer commented:
Just to be sure that I don't put nonsense in the release notes - would this be acceptable or should it include more details?

```
* Fixed incorrect timezone warnings in custom admin templates
  (:ticket:`25845`).
```

Also, should I squash my commits if I make further changes so that the pull request only contains one commit like now?

timgraham commented:
I would say "in custom admin templates that don't a data-... attribute in the body tag." Maybe the upgrade tip about adding it in custom templates could be mentioned in the 1.9 release too?

Yes, you can squash commits when updating.

Strayer commented:
I updated the PR with the discussed changes. The fix in DateTimeShortcuts.js is now way simpler. The check on `if (body.getAttribute(...)) { ... }` should be enough to reliably test if the body attribute is set:
```
"" ? true : false
false
null ? true : false
false
undefined ? true : false
false

"0" ? true : false
true
```

I removed the Selenium test and instead added another assert in the DateTimeShortcuts.test.js' init() test. I had problems getting the JS tests to run with grunt, but the test worked in the browser.

timgraham commented:
merged in 9af40f5df13801ffadcc5ded7440e4616123959f, thanks!
